libs/utils/trace.py

```diff
@@ -674,6 +677,23 @@ def _sanitize_CpuFrequency(self):
             return
         df = self._dfg_trace_event('cpu_frequency')
         clusters = self.platform['clusters']
+
+        # devlib introduces fake cpu_frequency events to make sure the
+        # frequency is known at the "beginning" of a trace
+        if self.hasEvents('cpu_frequency_devlib'):
+            devlib_freq = self._dfg_trace_event('cpu_frequency_devlib')
+            devlib_freq.rename(columns={'cpu_id':'cpu'}, inplace=True)
+            devlib_freq.rename(columns={'state':'frequency'}, inplace=True)
+            # Only use fake events if introduced before any cpufreq governor
+            # did, otherwise they may end up being interleaved with real events
+            # and make the the frequency coherency check fail
+            if len(devlib_freq) > 0:
+                first_devlib_freq = devlib_freq.iloc[:self.platform['cpus_count']]
+                if df.index[0] > first_devlib_freq.index[-1]:
+                    df = pd.concat([first_devlib_freq, df])
+            setattr(self.ftrace.cpu_frequency, 'data_frame', df)
```

Comment:
Well, most of the `_sanitize_*` methods are required to add custom columns and compute values to a TRAPpy table. Thus we are in the domain of mangling dataframes which is exactly what we do here.

The main issue for me to move certain bits into TRAPpy is that, sometimes we need platform data which are available only at the LISA side. Thus, in general I would avoid mixed approaches where we have somethings done in LISA others in TRAPpy.

IMHO, perhaps considering TRAPpy as a "raw" trace events parser while additional logic is added on top of LISA.